### DISCUSSION TITLE
Saving predictions in 3d_segmentation/unetr_btcv_segmentation_3d.ipynb as nifti format

### DISCUSSION THREAD

**Discussion by nicofutur8** [Project-MONAI/MONAI#3760] [General]
Hello everyone,

I am fairly new to MONAI and would like to save my UNETR predictions in nifti format. Is there any tutorial that can help me? I can't understand what transformations I have to do and how to use the SaveImage transformation.

Thank you very much for the help and I hope I have been clear with my doubt!

PD: i tried to replicate validation's transforms but i save a One-hot format (multichannel binary label) and i want a multilabel format (only a channel with 14 labels)

**Comment by Nic-Ma** [comment]
Hi @nicofutur8 ,

Thanks for your interest here.
I think you can refer to this tutorial: https://github.com/Project-MONAI/tutorials/blob/master/3d_segmentation/unetr_btcv_segmentation_3d.ipynb.
And for any further questions, I think @ahatamiz can help share more details.

Thanks in advance.

**Reply by nicofutur8** [reply]
Thanks for the fast answer @Nic-Ma. The problem with this tutorial is it cannot save prediction results in any format (only plot a result by matplotlib). I tried to save them by data transform `SaveImage` but i can't load the nifti file in Slicer3D (Slicer doesnt provide any information). I have loaded a prediction file in Mango and a label from training dataset and notice there are some changes in their headers and info: 
![label](https://user-images.githubusercontent.com/47743575/152193210-29ef09b1-167c-4755-8405-736da7fa6560.jpg)
![Captura de pantalla 2022-02-02 171652](https://user-images.githubusercontent.com/47743575/152193216-e41e6285-a572-4f3f-b005-1927f1fd1eb6.jpg)

The voxel dimensions doesn't have "unit type" on the pred one. Can be the reason of Slicer's error?

**Comment by ahatamiz** [comment]
Hi @nicofutur8 

Thanks for your interest in our work. I would recommend something like the following as an example:

```
import nibabel as nib
import numpy as np
for i, batch in enumerate(val_loader):
    val_inputs= batch["image"].cuda()
    affine = batch['image_meta_dict']['original_affine'][0].numpy()
    val_outputs = sliding_window_inference(val_inputs,
                                           (96, 96, 96),
                                           4,
                                           model,
                                           overlap=0.7)
    val_outputs = torch.softmax(val_outputs, 1).cpu().numpy()
    val_outputs = np.argmax(val_outputs, axis=1).astype(np.uint8)[0]
    nib.save(nib.Nifti1Image(val_outputs, affine), path_to_save)
```

Please also check the evaluation [script](https://github.com/Project-MONAI/research-contributions/blob/master/UNETR/BTCV/test.py) in UNETR repository. 

Thanks

**Reply by Nic-Ma** [reply]
Thanks @ahatamiz for example code.
And if you want to use MONAI APIs (Activation, AsDiscrete, SaveImage, etc.), please refer to:
https://github.com/Project-MONAI/tutorials/blob/master/3d_segmentation/torch/unet_evaluation_dict.py#L97

Thanks.

**Reply by nicofutur8** [reply]
Thanks for the code! I have come to a good solution with the `AsDiscrete` function and using `ITK` library. If you want, I can give you the code and add it to the tutorial, since inference is an important step in the workflow of artificial intelligence.

Good job with MONAI!

**Reply by Alex8211101** [reply]
Hello,I encountered a problem of origin mismatch when exporting data in nifti format.Can I learn this part of your code?Thank you very much!

**Reply by ahatamiz** [reply]
Hi @Alex8211101 

I recently added saving of segmentation masks in the original spacing. Please see the following:

[https://github.com/Project-MONAI/research-contributions/blob/main/SwinUNETR/BTCV/test.py#L106](https://github.com/Project-MONAI/research-contributions/blob/main/SwinUNETR/BTCV/test.py#L106)

Thanks
